service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+		currentVersionedTransition,
+	)
+
+	err := ms.applyUpdatesToStateMachineNodes(mutation.UpdatedSubStateMachines)
+	if err != nil {
+		return err
+	}
+
+	ms.approximateSize += ms.executionInfo.Size() - prevExecutionInfoSize
+	return nil
+}
+
+func (ms *MutableStateImpl) ApplySnapshot(
+	snapshot *persistencespb.WorkflowMutableState,
+) error {
+	prevExecutionInfoSize := ms.executionInfo.Size()
+	currentVersionedTransition := ms.currentVersionedTransition()
+
+	ms.applySignalRequestedIds(snapshot.SignalRequestedIds, snapshot.ExecutionInfo)
+	ms.applyTombstones(snapshot.ExecutionInfo.SubStateMachineTombstoneBatches, currentVersionedTransition)
+	ms.syncExecutionInfo(ms.executionInfo, snapshot.ExecutionInfo)
+	ms.syncSubStateMachinesByType(snapshot.ExecutionInfo.SubStateMachinesByType)
+
+	ms.approximateSize += snapshot.ExecutionState.Size() - ms.executionState.Size()
+	ms.executionState = snapshot.ExecutionState
+
+	ms.applyUpdatesToSubStateMachines(
+		snapshot.ActivityInfos,
+		snapshot.TimerInfos,
+		snapshot.ChildExecutionInfos,
+		snapshot.RequestCancelInfos,
+		snapshot.SignalInfos,
+		currentVersionedTransition,
+	)
+
+	ms.approximateSize += ms.executionInfo.Size() - prevExecutionInfoSize
+	return nil
+}
+
+func (ms *MutableStateImpl) applyUpdatesToSubStateMachines(
+	updatedActivityInfos map[int64]*persistencespb.ActivityInfo,
+	updatedTimerInfos map[string]*persistencespb.TimerInfo,
+	updatedChildExecutionInfos map[int64]*persistencespb.ChildExecutionInfo,
+	updatedRequestCancelInfos map[int64]*persistencespb.RequestCancelInfo,
+	updatedSignalInfos map[int64]*persistencespb.SignalInfo,
+	currentVersionedTransition *persistencespb.VersionedTransition,
+) {
+	applyUpdatesToSubStateMachine(ms, ms.pendingActivityInfoIDs, ms.updateActivityInfos, updatedActivityInfos, currentVersionedTransition, func(current, incoming *persistencespb.ActivityInfo) {
+		incoming.TimerTaskStatus = TimerTaskStatusNone
+	})
+
+	applyUpdatesToSubStateMachine(ms, ms.pendingTimerInfoIDs, ms.updateTimerInfos, updatedTimerInfos, currentVersionedTransition, func(current, incoming *persistencespb.TimerInfo) {
+		incoming.TaskStatus = TimerTaskStatusNone
+	})
+
+	applyUpdatesToSubStateMachine(ms, ms.pendingChildExecutionInfoIDs, ms.updateChildExecutionInfos, updatedChildExecutionInfos, currentVersionedTransition, func(current, incoming *persistencespb.ChildExecutionInfo) {
+		if current != nil {
+			incoming.Clock = current.Clock
+		}
+		incoming.CreateRequestId = uuid.New()
+	})
+
+	applyUpdatesToSubStateMachine(ms, ms.pendingRequestCancelInfoIDs, ms.updateRequestCancelInfos, updatedRequestCancelInfos, currentVersionedTransition, nil)
+	applyUpdatesToSubStateMachine(ms, ms.pendingSignalInfoIDs, ms.updateSignalInfos, updatedSignalInfos, currentVersionedTransition, nil)
+}
+
+func (ms *MutableStateImpl) applyUpdatesToStateMachineNodes(
+	nodeMutations []*persistencespb.WorkflowMutableStateMutation_StateMachineNodeMutation,
+) error {
+	root := ms.HSM()
```

Comment:
oh I was originally suggesting update ms.executionInfo.SubStateMachinesByType directly. But looks like this works too 👍